host/frontend/host-orchestrator/clients.go

```diff
@@ -0,0 +1,156 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+//     Unless required by applicable law or agreed to in writing, software
+//     distributed under the License is distributed on an "AS IS" BASIS,
+//     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//     See the License for the specific language governing permissions and
+//     limitations under the License.
+
+package main
+
+import (
+	"math/rand"
+	"strings"
+	"sync"
+	"time"
+)
+
+type Client interface {
+	// Send a message to the device
+	Send(msg interface{}) bool
+	// Provides an oportunity for the client to react to the device being disconnected
+	OnDeviceDisconnected()
+}
+
+// Implements the client interface using a websocket connection
+type WsClient struct {
+	ws *JsonWs
+}
+
+func NewWsClient(ws *JsonWs) *WsClient {
+	return &WsClient{ws: ws}
+}
+
+// From IClient
+func (c *WsClient) Send(msg interface{}) bool {
+	return c.ws.Send(msg)
+}
+func (c *WsClient) OnDeviceDisconnected() {
+	// Do nothing
+}
+
+// Implements the client interface using a polled connection
+type PolledClient struct {
+	// The polled connection id
+	id       string
+	device   *Device
+	messages []interface{}
+	// Synchronizes access to the messages list
+	msgMtx sync.Mutex
+	// The id given to this client by the device
+	clientId int
+	// A reference to the polled set to clean up when the device disconnects
+	polledSet *PolledSet
+}
+
+// From IClient
+func (c *PolledClient) Send(msg interface{}) bool {
+	c.msgMtx.Lock()
+	defer c.msgMtx.Unlock()
+	c.messages = append(c.messages, msg)
+	return true
+}
+func (c *PolledClient) OnDeviceDisconnected() {
+	// This object is useless when the device disconnects, release all resourcess associated to it
+	c.polledSet.Destroy(c.id)
+}
+
+// Sends a message to the device
+func (c *PolledClient) ToDevice(msg interface{}) bool {
+	return c.device.Send(msg)
+}
+
+// Gets count messages from the device, skipping the first start messages
```

Comment:
Are messages ever cleared? Can a long-running device pile up messages indefinitely?

Reply:
Messages are cleared when the device disconnects. In theory a device can pile up messages indefinitely, but devices will be connecting over a unix socket, so they'll be trusted. A malicious user can still create many clients and cause a device to send many messages to each, but that would still be the case if messages were cleared after read.
We could limit the number or clients per device and the number of messages, but what's a good limit?
We could also simply limit the memory consumption of the process (which is what I plan to do when making it start on boot)